libp2p/switch.nim

```diff
@@ -239,21 +237,35 @@ proc dial*(s: Switch,
   else:
     trace "Reusing existing connection"
 
+  await s.subscribeToPeer(peer)
+  result = conn
+
+proc connect*(s: Switch, peer: PeerInfo) {.async.} =
+  var conn = await s.internalConnect(peer)
```

Comment:
so sometimes this uses exceptions on error, and sometimes nil?

also, can it happen that upgrading connection fails for one transport / address but not another? seems plausible (say they can't agree on a cipher suite with noise so they fall back on secio)? 

in the case `upgradeOutgoing` raises, the `conn` object will have its peerinfo set, is that right? if it's always supposed to be set and dial creates the connection, why not pass it to dial? why does `upgradeOutgoing` sometimes return `nil`? how is that different from raising?

Reply:
what happens if someone calls both `dial` and `connect` in parallel?

Reply:
I can't decipher this questions. In connect, you wont go through the upgrade flow.

Reply:
Ok, had to read it a few times, I'll do my best to answer:

> also, can it happen that upgrading connection fails for one transport / address but not another? 

Right now we simply dial all addresses in sequence until one succeeds, once we have a connection we upgrade it. But this are different steps, we only ever have one outgoing connection per peer right now, so we only every upgrade one outgoing connection.

> seems plausible (say they can't agree on a cipher suite with noise so they fall back on secio)?

This isn't currently possible because multistream-select is super dumb. It doesn't define any sort of matching priority or selection order, it simply tries all protocols it has registered and picks the first one that succeeds. I know, it sounds ridiculous, but that's how it works and that's also why we need a much better protocol negotiation mechanism. So, right now, if you want to use one secure protocol, like Noise, you have to register only that protocol.

> in the case upgradeOutgoing raises, the conn object will have its peerinfo set, is that right?

Why would this be an issue? Also, the peerinfo gets updated during protocol negotiation.

>  if it's always supposed to be set and dial creates the connection, why not pass it to dial? 

Don't get this question.

> why does upgradeOutgoing sometimes return nil

Because the upgrade can fail for any number of reasons, for example, during the secure phase or identify phase, etc...